---
title: "SpideR Guide"
author: "AS"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---


# Sources of Documentation

```{r setup, include=FALSE}
knitr::opts_chunk$set(eval = FALSE)
```

This vignette aims to provide the most comprehensive piece of documentation on the SpideR package (NOTE: currently under construction ===*=== ).  
<br>
For a quick reference, please refer to Quick_SpideR vignette.  
<br>
Package documentation can also be viewed via help pages, i.e.  
```{r}
#Main page of package documentation
package?SpideR

# Documentation for individual functions
?searchForTerm
?searchForAccession
#etc.
```


# SpideR Overview

SpideR allows for search and retrieval of data from SRA and GEO databases.  
Its aim is to provide a more unified framework to access these two databases, integrate information between them and allow efficient and convenient retrieval of the data.  
It can be used in a variety of ways including tasks such as:  

- Searching for experiments that were done about X  
- Searching for experiments based on their accession number  
- Converting between accession numbers  

# Requirements

SpideR relies on the SRAdb and GEOmetadb packages. They provide large database files (at the time of writing about 40 GB in total).  
The user can specify the location of these files when running startSpideR(dir="Path_To_Folder_Of_Interest").  

<br>
In general, the requirements for running SpideR include:  

- R (we recommend to use it via RStudio)  
- dependency R packages (as per DESCRIPTION file), including:  
    * SRAdb  
    * GEOmetadb 
    * **NOTE:** these two need to be installed manually via Bioconductor  
- at least 40 GB of free disc space  


# Setup

1. Before proceeding, please ensure that you have the following packages installed (they are distributed via Bioconductor):

- SRAdb
- GEOmetadb

2. Install SpideR.

Under the development version ===*===, this is done by downloading the SpideR directory, opening a session in that directory and running:
```{r}
devtools::load_all()
```


3. Initialise SpideR.

This has to be done every time you want to use SpideR in a new session.

```{r}
startSpideR(dir=getwd()) # dir specifies the location where database files will be stored
```


It may take a considerable amount of time the first time it is run. This is due to the need for downloading large database files and creating an extra custom database. 


However, once these are in place, the above command should take a matter of seconds or less.  

NOTE: it is necessary to download large database files in order to use any of the SpideR functions. However, it is not strictly necessary to have the most up-to-date files; the SpideR will work even if startSpideR generates warnings to get newer database files.



# SpideR by Functions

## Brief Overview

1. searchForTerm - search for (unknown) experiments matching search terms of interest
2. searchForAccession - search for (known) experiments which match the accession numbers
3. convertAccession - convert between SRA and/or GEO accessions
4. reproduceResults / rerunQuery (function name will be changed ===*=== ) - run again a SpideR query from the past (NB: the results might differ if the database changed)
5. getDatabaseInformation - get basic information about the databases


## Search for (unknown) experiments - by search terms (searchForTerm)

```{r}
searchForTerm(library_strategy, gene=NULL, antibody=NULL, cell_type=NULL, treatment=NULL, species=NULL, platform=NULL, secondary_library_strategy=NULL) 
```

REQUIRED ARGUMENTS: library_strategy AND at least one of gene, antibody, cell_type or treatment 
OPTIONAL ARGUMENTS: species, platform, secondary_library_strategy

###Argument format
* library_strategy - one of the standardised entries, such as: "ChIP-Seq", "RNA-Seq", "HiC" etc. ===*=== [to do: provide a full list]
* gene* - a name of the gene. It is recommended that a vector with synonyms is used, for example gene=c("STAT1", "STAT-1", "STAT91", "STAT-91", "ISGF-3", "ISGF3") instead of gene="STAT1". However, adding multiple synonyms can extend the query time
* antibody* - a name of the antibody. Similarly, a vector with synonyms can be used
* treatment* - information on treatment ===*=== [to do: double check if really can be a vector]
* species* - code for species, for example species=c("9606", "10090") for human and mouse entries
* platform* - sequencing platform, for example platform-"ILLUMINA"
* secondary_library_strategy - an extra library_strategy that will be used to filter the entries from the same studies as the matching entries.  

An asterisk (\*) denotes arguments which can be vectors. In such case an entry will be returned if there is a match to ANY ONE of the vector elements
NOTE:

* The primary use for the antibody argument is for relevant library strategies, such as ChIP-Seq. ===*=== [to do: add a warning if otherwise]
* Some studies fill their library_strategy as "OTHER", so if more results are needed, rerun the same query with library_strategy="OTHER"
* Secondary_library_strategy can be used to find entries within the same studies, which have a library_strategy of interest.  
For example, if searchForTerm(library_strategy="ChIP-Seq", antibody="STAT1") yielded entries A, B, C, D, then searchForTerm(library_strategy="ChIP-Seq", antibody="STAT1", secondary_library_strategy="RNA-Seq") would produce the same results AND on top of that return any RNA-Seq entries within the same studies as A, B, C, D.

###Sample query

```{r}
searchForTerm(library_strategy="ChIP-Seq", antibody="STAT1") # This will search for all entries which have "STAT1" in antibody field
```


NOTE:

* Not all submissions have 'antibody' information specifically. When that is the case, other fields are searched, which might lead to false positives




###Output files
The searchForTerm function outputs a number of files. Depending on the application, some or all of them might be useful for the user.  
The current naming system includes the main keywords of the query as well as date and time. The middle part of the file name depends on the output file type.   

The current file name structure is as follows:  
(Query summary)\_(file type)\_(date)_(time).extension

The legend for file types:  

- db - database extract (i.e. lists of relevant samples with associated information)
- sm - sample sheet
- CALL - function call
- PAR - parameters of the function call (can be used to re-run the query with the reproduceResults function)  


Depending on the contents the database extracts/sample sheets will have the following additional attributes:

- tar - only runs of interest (and any potential inputs/controls)
- all - all runs within a study (which belong to the library strategy of interest)
- 2ry - all runs within a study filtered by the secondary library strategy. This file is only output if secondary library strategy has been specified


For example, the query 
```{r}
searchForTerm(library_strategy = "ChIP-Seq", antibody = "STAT1", secondary_library_strategy = "RNA-Seq")
```



will generate the following set of output files: 

* ChIP_STAT1_2RNA_CALL_180213-110628.Rda 
* ChIP_STAT1_2RNA_PAR_180213-110627.tab 
* ChIP_STAT1_2RNA_tar_db_180213-110816.csv
* ChIP_STAT1_2RNA_tar_db_180213-110816.Rda
* ChIP_STAT1_2RNA_tar_sm_180213-110816.csv
* ChIP_STAT1_2RNA_tar_sm_180213-110816.Rda
* ChIP_STAT1_2RNA_all_db_180213-110817.csv 
* ChIP_STAT1_2RNA_all_db_180213-110817.Rda 
* ChIP_STAT1_2RNA_all_sm_180213-110816.csv
* ChIP_STAT1_2RNA_all_sm_180213-110816.Rda
* ChIP_STAT1_2RNA_2ry_db_180213-110817.csv
* ChIP_STAT1_2RNA_2ry_db_180213-110817.Rda
* ChIP_STAT1_2RNA_2ry_sm_180213-110817.csv
* ChIP_STAT1_2RNA_2ry_sm_180213-110817.Rda


These files can be subdivided depending on their content:


- Files for reproducing the query:  

    * ChIP_STAT1_2RNA_CALL_180213-110628.Rda  - R object with the function call  
    * ChIP_STAT1_2RNA_PAR_180213-110627.tab - tab file with parameters of the function call  
  
- Files with ChIP-Seq samples of interest (and any associated inputs); the set of studies to which these belong is used for the two other sets of outputs, as outlined below:

    * ChIP_STAT1_2RNA_tar_db_180213-110816.csv  - database extract
    * ChIP_STAT1_2RNA_tar_db_180213-110816.Rda  - database extract
    * ChIP_STAT1_2RNA_tar_sm_180213-110816.csv  - sample sheet
    * ChIP_STAT1_2RNA_tar_sm_180213-110816.Rda - sample sheet
  
- Files with all ChIP-Seq samples from the same studies as the samples in ...\_tar\_... file (any associated inputs will naturally be included):

    * ChIP_STAT1_2RNA_all_db_180213-110817.csv  - database extract
    * ChIP_STAT1_2RNA_all_db_180213-110817.Rda  - database extract
    * ChIP_STAT1_2RNA_all_sm_180213-110816.csv  - sample sheet
    * ChIP_STAT1_2RNA_all_sm_180213-110816.Rda  - sample sheet
    
- Files with all RNA-Seq samples from the same studies as the samples in ...\_tar\_... file

    * ChIP_STAT1_2RNA_2ry_db_180213-110817.csv  - database extract 
    * ChIP_STAT1_2RNA_2ry_db_180213-110817.Rda  - database extract
    * ChIP_STAT1_2RNA_2ry_sm_180213-110817.csv  - sample sheet
    * ChIP_STAT1_2RNA_2ry_sm_180213-110817.Rda  - sample sheet


Where:  

- ChIP_STAT1_2RNA containS compressed information on the query  
- central part of the file name refers to its contents/type  
- 180213-110817 is the date and time of file creation  




## Search for (known) experiments - by accession numbers (searchForAccession)

**TASK:** *"I would like to retrieve a handy list of samples from a study I already know."*  
```{r, eval = FALSE}
searchForAccession("SRP010068")
```









To be included: ===*===

- what is this function for
- argument requirements
- explanation on the format of the arguments (vector vs not, matches)
- info on outputs and file naming
- advice on what to do to maximise results (e.g. flexible search, other library strategy)




-------



https://www.ncbi.nlm.nih.gov/taxonomy 